drivers/SmartThings/matter-rvc/src/init.lua

```diff
@@ -28,213 +25,232 @@ if version.api < 10 then
+    local component = device.profile.components["runMode"]
+    local event = capabilities.mode.supportedArguments({}, {visibility = {displayed = false}})
+    device:emit_component_event(component, event)
+    -- Set cleanMode to empty
+    component = device.profile.components["cleanMode"]
+    device:emit_component_event(component, event)
+    return
   end
-  return false
-end
 
-local function get_labels_of_supported_modes_filter_by_current_mode(device, field_prefix, idle_mode_tag, current_mode)
-  local labels = get_field_labels_of_supported_modes(device, field_prefix)
-  local is_idle_current_mode = false
-  for i, label in ipairs(labels) do
-    if label == current_mode then
-      is_idle_current_mode = is_idle_mode(device, field_prefix, i, idle_mode_tag)
+  -- Get the tag of the current run mode
+  local current_tag = 0xFFFF
+  local supported_run_modes = device:get_field(RUN_MODE_SUPPORTED_MODES) or {}
+  for i, mode in ipairs(supported_run_modes) do
+    if mode[1] == current_run_mode then
+      current_tag = mode[2]
       break
     end
   end
+  if current_tag == 0xFFFF then
+    return
+  end
 
-  if is_idle_current_mode then
-    return labels
+  -- Check whether non-idle mode can be selected or not
+  local op_state = capabilities.robotCleanerOperatingState.operatingState
+  local can_be_selected_as_non_idle = 0
+  if current_tag == clusters.RvcRunMode.types.ModeTag.IDLE and
+    (operating_state == op_state.stopped.NAME or operating_state == op_state.paused.NAME or
+     operating_state == op_state.docked.NAME or operating_state == op_state.charging.NAME) then
+      can_be_selected_as_non_idle = 1
   end
 
-  local filtered_labels = {}
-  table.insert(filtered_labels, current_mode)
-  for i, label in ipairs(labels) do
-    if is_idle_mode(device, field_prefix, i, idle_mode_tag) then
-      table.insert(filtered_labels, label)
+  -- Set supported run arguments
+  local supported_arguments = {} -- For generic plugin
+  for i, mode in ipairs(supported_run_modes) do
+    if mode[2] == clusters.RvcRunMode.types.ModeTag.IDLE or can_be_selected_as_non_idle == 1 then
+      table.insert(supported_arguments, mode[1])
     end
   end
-  return filtered_labels
+
+  -- Send event to set supported run arguments
+  local component = device.profile.components["runMode"]
+  local event = capabilities.mode.supportedArguments(supported_arguments, {visibility = {displayed = false}})
+  device:emit_component_event(component, event)
+
+  -- Set supported clean arguments
+  local supported_clean_modes = device:get_field(CLEAN_MODE_SUPPORTED_MODES) or {}
+  supported_arguments = {}
+  for i, mode in ipairs(supported_clean_modes) do
+    table.insert(supported_arguments, mode)
+  end
+
+  -- Send event to set supported clean modes
+  local component = device.profile.components["cleanMode"]
+  local event = capabilities.mode.supportedArguments(supported_arguments, {visibility = {displayed = false}})
+  device:emit_component_event(component, event)
```

Comment:
```suggestion
  local component = device.profile.components["cleanMode"]
  if component ~= nil then
    local event = capabilities.mode.supportedArguments(supported_arguments, {visibility = {displayed = false}})
    device:emit_component_event(component, event)
  end
```
In case cleanMode isn't supported and the device is using the `rvc.yml` profile.